### ISSUE TITLE
[Question] Frozen parameter equals true  for distributions dont skip parameters calculations

### ISSUE THREAD

**Issue by oniram-reach** [jmschrei/pomegranate#891]
I would like to run my Hidden Markov model with GammaDistribution and fixed parameters.
But when printed model returned by from_samples, I see that the distribution parameters have been updated and frozen atribute is false.

Steps to reproduce:

`data =  np.array([[22.10,21.73,21.78,21.51,21.37,21.16,20.72,20.50]])`
`number_hidden_states = 2`
`d = GammaDistribution(5, 2,True)`

`dists = [d] * number_hidden_states`

`model = HiddenMarkovModel.from_samples(dists, number_hidden_states, 
			data, max_iterations=25)`

When print dists its Ok

`    "class" : "Distribution",
    "name" : "GammaDistribution",
    "parameters" : [
        5.0,
        2.0
    ],
    "frozen" : true
}, {
    "class" : "Distribution",
    "name" : "GammaDistribution",
    "parameters" : [
        5.0,
        2.0
    ],
    "frozen" : true
}]
`

but when print model the frozen is false and paramters alpha/beta was updated to [9112.18402722927, 418.63449037403245]:

`None:{
    "class" : "State",
    "distribution" : {
        "class" : "Distribution",
        "name" : "IndependentComponentsDistribution",
        "parameters" : [
            [
                {
                    "class" : "Distribution",
                    "name" : "GammaDistribution",
                    "parameters" : [
                        9112.18402722927,
                        418.63449037403245
                    ],
                    "frozen" : false
                }
            ],
            [
                1.0
            ]
        ],
        "frozen" : false
    },
    "name" : "s0",
    "weight" : 1.0
}`

I doing something wrong to run my model with the parameters I defined?

**Comment by jmschrei** [comment]
That's weird. https://github.com/jmschrei/pomegranate/blob/master/pomegranate/distributions/GammaDistribution.pyx#L157 should prevent it from being updated. The shape of your data is a bit weird, though. Your data object is shaped such that you have one example with 8 dimensions, but then you specify your distribution to have two dimensions by being a list with two univariate distributions. If neither of those work, try passing `distribution_inertia=1.0` into `from_samples`.

**Comment by oniram-reach** [comment]
Thanks for reply!

I change my title for Question label.

it seems that I have not clearly defined the shape of my data array then.
In the example I put only a part of it, but in fact my data is a list of about 250 items that represent stock prices. And I used np.array just to avoid a mistake I made when passing as a list.
Do you indicate any reference where I can learn how my data should look in relation to the Gamma dimension and distribution that I want to evaluate?

Obs: passing distribution_inertia=1.0 into from_samples continue updating gamma parameters

**Comment by jmschrei** [comment]
For an HMM, your data should be a list of examples. Each example should be of shape (sequence_length, dimensionality) if it's multivariate or just (sequence_length,) if univariate. I wasn't looking at what you did carefully enough. If you intended to pass in a single example that is univariate and has 8 observations, you did the correct thing. 

I looked into what was going on, and the problem arises because pomegranate treats the distributions that you pass in as classes rather than as instantiated objects. If you print out `d` after training you can see that it has not been changed at all. Basically, internally, it's using the GammaDistribution class to create new objects that then get fit, without also copying the parameters that you passed in because it's not expecting those. The expected use is `HiddenMarkovModel.from_samples(GammaDistribution, number_hidden_states, data, max_iterations=25)`.

To get the usage you'd like you can either manually construct the HMM using `add_states` and `add_edge` commands, or you can use the `from_matrix` method where you pass in a list of distribution objects (not classes) as well as a transition matrix and a vector of start probabilities. The code would look something like this:

```python
import numpy as np
from pomegranate import *

data = np.array([[22.10,21.73,21.78,21.51,21.37,21.16,20.72,20.50]])
number_hidden_states = 2
d = GammaDistribution(5, 2,True)

dists = [d] * number_hidden_states
starts = numpy.ones(number_hidden_states, dtype='float32') / number_hidden_states
transitions = numpy.ones((number_hidden_states, number_hidden_states), dtype='float32') / number_hidden_states

model = HiddenMarkovModel.from_matrix(transitions, dists, starts)
model.fit(data, max_iterations=25, verbose=True)

print(d)
print(model.states[0].distribution)
```

You'll see that when calling `fit` that the `frozen=True` in the distribution is respected. You'll also see that you don't learn anything. This is because both of your distributions have exactly the same parameters. You need to change the distribution objects to have different parameters in order to fit a transition matrix.

**Comment by oniram-reach** [comment]
Thanks!
